File: bf.py
#!/usr/bin/env python3

from sys import argv, stdin, stderr
cells = bytearray(30000) # Memory (30kb)
dp = 0 # Data pointer
    
try:
    with open(argv[1], "r") as f:
        code = f.read()
except IndexError:
    print("Enter code here:")
    code = stdin.read()
    print("")

stack = [] # Bracket nest stack
jump = [None]*len(code)
ip = 0 # Instruction pointer

if code.count("[") != code.count("]"):
    stderr.write("Error: Brackets are unbalanced\n")
    exit(1)

# totally not taken from https://stackoverflow.com/a/3041005
for i,o in enumerate(code):
    if o=='[':
        stack.append(i)
    elif o==']':
        try:
            jump[i] = stack.pop()
        except IndexError:
            stderr.write("Error: Brackets are unbalanced\n")
            exit(1)
        jump[jump[i]] = i

while ip < len(code):
    match code[ip]:
        case "+":
            cells[dp] = (cells[dp] + 1) % 256
        case "-":
            cells[dp] = (cells[dp] - 1) % 256
        case ">":
            dp+=1
            if dp < -1 or dp > 29999:
                dp -= 30000
        case "<":
            dp-=1
            if dp < -1 or dp > 29999:
                dp += 30000
        case ".":
            print(chr(cells[dp]), end="")
        case ",":
            theinput = input("\nInput: ")
            if theinput != "":
                 cells[dp] = ord(theinput[0])
            else:
                cells[dp] = 0
        case "[":
            if not cells[dp]:
                ip = jump[ip]
        case "]":
            if cells[dp]:
                ip = jump[ip]
                continue
     ip+=1

print("")

File: bf.lua
--[[Copyright 2026 rock3tsprocket

Redistribution and use in source and binary forms, with or without modification, are permitted
provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this list of conditions and
the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions
and the following disclaimer in the documentation and/or other materials provided with the
distribution.

3. Neither the name of the copyright holder nor the names of its contributors may be used to
endorse or promote products derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS “AS IS”
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.]]

if arg[1] == nil then
	print("No file was specified")
	os.exit(1)
end
f = assert(io.open(arg[1], "r"))
code = f:read("*all")
assert(f:close())

codelen = string.len(code) -- Length of code
dp = 0 -- Data pointer
ip = 0 -- Instruction pointer

--Doing cell things
cells = {}
for i=0,29999 do
	cells[i] = 0
end

-- Jump table
stack = {}
stacksize = 0
jumptable = {}
for i=0, codelen do
	jumptable[i] = 0
end
for i=1, codelen do
	if string.sub(code, i, i) == "[" then
		table.insert(stack, i)
		stacksize = stacksize + 1
	elseif string.sub(code, i, i) == "]" then
		jumptable[i] = table.remove(stack, stacksize)
		stacksize = stacksize - 1
		jumptable[jumptable[i]] = i
	end
end

-- Main loop
while ip <= codelen do
	if string.sub(code, ip, ip) == "+" then
		cells[dp] = (cells[dp] + 1) % 256
	elseif string.sub(code, ip, ip) == "-" then
		cells[dp] = (cells[dp] - 1) % 256
		if cells[dp] <= -1 then
			cells[dp]=cells[dp]+256
		end
	elseif string.sub(code, ip, ip) == "." then
		io.stdout:write(string.char(cells[dp]))
        io.stdout:flush()
	elseif string.sub(code, ip, ip) == "," then
		io.stdout:write("Input: ")
		input = io.stdin:read()
		if input ~= "" then
			cells[dp] = string.byte(string.sub(input, 1, 1))
		else
			cells[dp] = 0
		end

	elseif string.sub(code, ip, ip) == ">" then
		dp = (dp + 1) % 30000
	elseif string.sub(code, ip, ip) == "<" then
		dp = (dp - 1) % 30000
	elseif string.sub(code, ip, ip) == "[" and cells[dp] == 0 then
		ip = jumptable[ip]
	elseif string.sub(code, ip, ip) == "]" and cells[dp] ~= 0 then
		ip = jumptable[ip]
	end

	ip = ip + 1
end
